### DISCUSSION TITLE
Factory Recovery Feature Implementation

### DISCUSSION THREAD

**Discussion by AGlass0fMilk** [mcu-tools/mcuboot#1229] [Ideas]
Hi everyone,

One feature I've been thinking about implementing using MCUboot is having a factory recovery image. This would be a separate update binary that simply reverts the installed firmware to a "known-good" version that was installed at the factory. This could be implemented a number of different ways:

1.) Outside of MCUboot, the secondary flash area provided to MCUboot could be retargeted to the factory recovery image based on user input
2.) Using MCUboot's multiple flash area features

The concern I have is that, in my configuration, MCUboot will be performing swap updates. In this case, I do not want MCUboot to swap the old image into the secondary slot.

I suppose this could be implemented at the flash driver level (disallow writes to the factory partition) but I'm not sure how MCUboot would handle this.

Does anyone think this could be a useful feature? Should we look into adding such support to MCUboot?

**Comment by d3zd3z** [comment]
You might want to look at the bootstrap feature, which is intended to provide the copy upgrade for the initial image. #207 was around to have the code use copy instead of swap when the image was marked as always for upgrade. It might be useful to revive that.

You will have to address at least the configuration issues with anti-rollback, and that current images are likely to be newer than the recovery image.

One possibility, is that normal mode may just be overwrite only, because it is only needed for certain types of recovery, which you would have with the recovery image.
